gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+            this.recordFields = recordFields;
+            this.constructor = constructor;
+            for (int i = 0; i < recordFields.length; i++) {
+                RecordField recordField = recordFields[i];
+                for (String name : recordField.deSerializedNames) {
+                    Integer prevIndex = recordFieldIndex.put(name, i);
+                    if (prevIndex != null) {
+                        throw new IllegalArgumentException("Both [" + recordFields[prevIndex] + "]"
+                                + " and [" + recordField + "] can be read from the same name [" + name + "]");
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void write(JsonWriter out, T value) throws IOException {
+            if (value == null) {
+                out.nullValue();
+                return;
+            }
+            out.beginObject();
+            for (RecordField recordField : recordFields) {
+                recordField.write(out, value);
+            }
+            out.endObject();
+        }
+
+        @Override
+        public T read(JsonReader in) throws IOException {
+            if (in.peek() == JsonToken.NULL) {
+                in.nextNull();
+                return null;
+            }
+            else if (in.peek() != JsonToken.BEGIN_OBJECT) {
+                throw new DeSerializeException("Expecting null or begin object at [" + in.getPath() + "]");
+            }
+            in.beginObject();
+            // This array will hold the value for each json field in the src object that we find, deserialized
+            // using a RecordField, that will also respect all Gson annotations. For fields that are not deserialized,
+            // the default null value will be used.
+            Object[] values = new Object[recordFields.length];
+            nextField:
+            while (in.peek() != JsonToken.END_OBJECT) {
+                String fieldName = in.nextName();
+                // Check if we know which field this name should deserialize to. If we do not find a matching index,
+                // then this field is one that is not present in our record, and ignored. We do not treat unknown
+                // object fields as an error, because the version of the class we have might differ from the source.
+                // -stun.
+                Integer fieldIndex = recordFieldIndex.get(fieldName);
+                if (fieldIndex != null) {
+                    values[fieldIndex] = recordFields[fieldIndex].read(in);
+                }
+                else {
+                    // Be sure to skip values we do not consume, otherwise the state of the parser will be wrong.
+                    in.skipValue();
+                }
+            }
+            in.endObject();
+            // At this point, we have not verified that all fields are present, we treat absent values as null
```

Comment:
Not sure about this. If any of the missing fields have primitive type then we will get a NullPointerException. It's probably better to throw an exception always rather than just sometimes? Or we need some more logic to supply a zero value for the missing fields.

Reply:
Fixed to initialize some default values, to ensure we have non-null values for primitives.